cpp/src_prims/sparse/linalg/symmetrize.cuh

```diff
@@ -0,0 +1,306 @@
+ * @param n: Number of rows
+ * @param k: Number of n_neighbors
+ * @param row_sizes: Input empty row sum 1 array(n)
+ * @param row_sizes2: Input empty row sum 2 array(n) for faster reduction
+ */
+template <typename value_idx = int64_t, typename value_t = float>
+__global__ static void symmetric_find_size(const value_t *restrict data,
+                                           const value_idx *restrict indices,
+                                           const value_idx n, const int k,
+                                           value_idx *restrict row_sizes,
+                                           value_idx *restrict row_sizes2) {
+  const auto row = blockIdx.x * blockDim.x + threadIdx.x;  // for every row
+  const auto j =
+    blockIdx.y * blockDim.y + threadIdx.y;  // for every item in row
+  if (row >= n || j >= k) return;
+
+  const auto col = indices[row * k + j];
+  if (j % 2)
+    atomicAdd(&row_sizes[col], value_idx(1));
+  else
+    atomicAdd(&row_sizes2[col], value_idx(1));
+}
+
+/**
+ * @brief Reduce sum(row_sizes) + k
+ * Reduction for symmetric_find_size kernel. Allows algo to be faster.
+ *
+ * @param n: Number of rows
+ * @param k: Number of n_neighbors
+ * @param row_sizes: Input row sum 1 array(n)
+ * @param row_sizes2: Input row sum 2 array(n) for faster reduction
+ */
+template <typename value_idx>
+__global__ static void reduce_find_size(const value_idx n, const int k,
+                                        value_idx *restrict row_sizes,
+                                        const value_idx *restrict row_sizes2) {
+  const auto i = (blockIdx.x * blockDim.x) + threadIdx.x;
+  if (i >= n) return;
+  row_sizes[i] += (row_sizes2[i] + k);
+}
+
+/**
+ * @brief Perform data + data.T operation.
+ * Can only run once row_sizes from the CSR matrix of data + data.T has been
+ * determined.
+ *
+ * @param edges: Input row sum array(n) after reduction
+ * @param data: Input knn distances(n, k)
+ * @param indices: Input knn indices(n, k)
+ * @param VAL: Output values for data + data.T
+ * @param COL: Output column indices for data + data.T
+ * @param ROW: Output row indices for data + data.T
+ * @param n: Number of rows
+ * @param k: Number of n_neighbors
+ */
+template <typename value_idx = int64_t, typename value_t = float>
+__global__ static void symmetric_sum(value_idx *restrict edges,
+                                     const value_t *restrict data,
+                                     const value_idx *restrict indices,
+                                     value_t *restrict VAL,
+                                     value_idx *restrict COL,
+                                     value_idx *restrict ROW, const value_idx n,
+                                     const int k) {
+  const int row = blockIdx.x * blockDim.x + threadIdx.x;  // for every row
+  const int j = blockIdx.y * blockDim.y + threadIdx.y;  // for every item in row
+  if (row >= n || j >= k) return;
+
+  const int col = indices[row * k + j];
```

Comment:
`indices` is a pointer to `value_idx` which is by default `int64_t` so this might fail. `auto` or `value_idx` will both be valid here. Same goes for variables `row` and `j` above and `original` and `transpose` below

Reply:
Ah, great catch! I also needed to update the `coo_symmetrize` prim.